[AOTInductor] Switch ProxyExecutor to use AtenTensorHandle

SherlockNoMad opened the pull request:
Summary: Switch ProxyExecutor to use AtenTensorHandle.

Test Plan: E2E Test

Differential Revision: D49471659




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

yifuwang commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -147,6 +147,13 @@ AOTITorchError aoti_torch_empty_strided(
 
+AOTITorchError aoti_torch_new_tensor(AtenTensorHandle* ret) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    at::Tensor* out_tensor = new at::Tensor();
```
Who is responsible for deallocating this?

SherlockNoMad replied:
In the codege, there will be a `create_raii_tensor_handle_for_temp` call latter, which binds a deleter for this.

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -141,6 +141,11 @@ AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_empty_strided(
 
+// This function will create a new uninitialized tensor object
+// and its pointer is returned through *out.
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_new_tensor(
```
I suggest to add `uninitialized` in the name to make that explicit.

desertfire reviewed (approved):
LGTM. You need to rebase to the latest code, and adapt to the API changes in RAIIAtenTensorHandle.

chenyang78 reviewed (approved):
Not necessarily in this PR, but can we come up with a test. Particularly, since we changed  fill_output_arg to use correct APIs, we seems to be able to run some test with extern kernels?

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1711,8 +1719,12 @@ def fill_args(arg, arg_type):
+                self.writeline(
+                    f"AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_new_uninitialized_tensor(&{arg}_handle));"
+                )
+                self.writeline(f"RAIIAtenTensorHandle {arg}({arg}_handle);")
```
Hmm, wondering if we need to guard these lines with ```config.aot_inductor.abi_compatible```?

SherlockNoMad replied:
no, `generate_extern_kernel_args_decl_if_needed` is only used in the fbcode path, which always need abi=true.

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1813,7 +1825,9 @@ def generate_extern_kernel_alloc_and_find_schema_if_needed_fbcode(
         tensor_call_args_str = ", ".join(tensor_call_args)
-        self.writeline(f"void* {tensor_args_var}[] = {{{tensor_call_args_str}}};")
+        self.writeline(
+            f"AtenTensorHandle {tensor_args_var}[] = {{{tensor_call_args_str}}};"
```
Same here - do we need to guard ```AtenTensorHandle``` with ```config.aot_inductor.abi_compatible```?

SherlockNoMad replied:
ditto. this is in fbcode path.

SherlockNoMad committed:
[PT2 Inference] Enable ProxyExecutor with Runtime (#109748)

Summary:

Switch ProxyExecutor to use AtenTensorHandle.

bypass-github-pytorch-ci-checks
OSS CI has a irrelevant failure.

Test Plan: E2E Test

Reviewed By: yifuwang

Differential Revision: D49471659

huydhn commented:
@pytorchbot drci

(Please ignore this, I'm testing Dr.CI)
